{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seqpandas as spd \n",
    "from seqpandas import pathing\n",
    "from Bio import SeqIO\n",
    "import gzip\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sample files used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gbk = 'tests/test-data/GCF_000013425.1.gbk.gz'\n",
    "fasta = 'tests/test-data/random_sequences.fasta.gz'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fasta to Pandas DataFrame example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_seq</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>dbxrefs</th>\n",
       "      <th>annotations</th>\n",
       "      <th>_per_letter_annotations</th>\n",
       "      <th>features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(C, G, A, T, A, T, T, C, G, A, T, C, C, G, C, ...</td>\n",
       "      <td>FWIRNKE01DKIF6</td>\n",
       "      <td>FWIRNKE01DKIF6</td>\n",
       "      <td>FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 ...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(G, C, G, A, G, C, A, G, C, A, A, T, C, A, T, ...</td>\n",
       "      <td>FWIRNKE01CDBE3</td>\n",
       "      <td>FWIRNKE01CDBE3</td>\n",
       "      <td>FWIRNKE01CDBE3 rank=0000320 x=854.0 y=2685.0 l...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(C, G, A, G, C, A, G, C, A, C, A, T, C, A, T, ...</td>\n",
       "      <td>FWIRNKE01BKZJJ</td>\n",
       "      <td>FWIRNKE01BKZJJ</td>\n",
       "      <td>FWIRNKE01BKZJJ rank=0000535 x=531.0 y=3933.0 l...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                _seq              id  \\\n",
       "0  (C, G, A, T, A, T, T, C, G, A, T, C, C, G, C, ...  FWIRNKE01DKIF6   \n",
       "1  (G, C, G, A, G, C, A, G, C, A, A, T, C, A, T, ...  FWIRNKE01CDBE3   \n",
       "2  (C, G, A, G, C, A, G, C, A, C, A, T, C, A, T, ...  FWIRNKE01BKZJJ   \n",
       "\n",
       "             name                                        description dbxrefs  \\\n",
       "0  FWIRNKE01DKIF6  FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 ...      []   \n",
       "1  FWIRNKE01CDBE3  FWIRNKE01CDBE3 rank=0000320 x=854.0 y=2685.0 l...      []   \n",
       "2  FWIRNKE01BKZJJ  FWIRNKE01BKZJJ rank=0000535 x=531.0 y=3933.0 l...      []   \n",
       "\n",
       "  annotations _per_letter_annotations features  \n",
       "0          {}                      {}       []  \n",
       "1          {}                      {}       []  \n",
       "2          {}                      {}       []  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_df = spd.read_seq(fasta, format='fasta')\n",
    "fasta_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# You can also import seqrecords directly from BioPythons parsing output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_seq</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>dbxrefs</th>\n",
       "      <th>annotations</th>\n",
       "      <th>_per_letter_annotations</th>\n",
       "      <th>features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(C, G, A, T, A, T, T, C, G, A, T, C, C, G, C, ...</td>\n",
       "      <td>FWIRNKE01DKIF6</td>\n",
       "      <td>FWIRNKE01DKIF6</td>\n",
       "      <td>FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 ...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(G, C, G, A, G, C, A, G, C, A, A, T, C, A, T, ...</td>\n",
       "      <td>FWIRNKE01CDBE3</td>\n",
       "      <td>FWIRNKE01CDBE3</td>\n",
       "      <td>FWIRNKE01CDBE3 rank=0000320 x=854.0 y=2685.0 l...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(C, G, A, G, C, A, G, C, A, C, A, T, C, A, T, ...</td>\n",
       "      <td>FWIRNKE01BKZJJ</td>\n",
       "      <td>FWIRNKE01BKZJJ</td>\n",
       "      <td>FWIRNKE01BKZJJ rank=0000535 x=531.0 y=3933.0 l...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                _seq              id  \\\n",
       "0  (C, G, A, T, A, T, T, C, G, A, T, C, C, G, C, ...  FWIRNKE01DKIF6   \n",
       "1  (G, C, G, A, G, C, A, G, C, A, A, T, C, A, T, ...  FWIRNKE01CDBE3   \n",
       "2  (C, G, A, G, C, A, G, C, A, C, A, T, C, A, T, ...  FWIRNKE01BKZJJ   \n",
       "\n",
       "             name                                        description dbxrefs  \\\n",
       "0  FWIRNKE01DKIF6  FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 ...      []   \n",
       "1  FWIRNKE01CDBE3  FWIRNKE01CDBE3 rank=0000320 x=854.0 y=2685.0 l...      []   \n",
       "2  FWIRNKE01BKZJJ  FWIRNKE01BKZJJ rank=0000535 x=531.0 y=3933.0 l...      []   \n",
       "\n",
       "  annotations _per_letter_annotations features  \n",
       "0          {}                      {}       []  \n",
       "1          {}                      {}       []  \n",
       "2          {}                      {}       []  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with gzip.open(fasta, \"rt\") as handle:\n",
    "    seqrecords = SeqIO.parse(handle, format='fasta')\n",
    "    fasta_df = spd.BioDataFrame.from_seqrecords(seqrecords)\n",
    "fasta_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Search description with little effort!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_seq</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>dbxrefs</th>\n",
       "      <th>annotations</th>\n",
       "      <th>_per_letter_annotations</th>\n",
       "      <th>features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(C, G, A, T, A, T, T, C, G, A, T, C, C, G, C, ...</td>\n",
       "      <td>FWIRNKE01DKIF6</td>\n",
       "      <td>FWIRNKE01DKIF6</td>\n",
       "      <td>FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 ...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                _seq              id  \\\n",
       "0  (C, G, A, T, A, T, T, C, G, A, T, C, C, G, C, ...  FWIRNKE01DKIF6   \n",
       "\n",
       "             name                                        description dbxrefs  \\\n",
       "0  FWIRNKE01DKIF6  FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 ...      []   \n",
       "\n",
       "  annotations _per_letter_annotations features  \n",
       "0          {}                      {}       []  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# I want all ranks of a certain number \n",
    "fasta_df[fasta_df['description'].str.contains('rank=0000177')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Better view of the layout for a single row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'_seq': Seq('CGATATTCGATCCGCATCGCTGCCCTACCCGTGGAGTGCCTCCCTCGGNGCAG'),\n",
       "  'id': 'FWIRNKE01DKIF6',\n",
       "  'name': 'FWIRNKE01DKIF6',\n",
       "  'description': 'FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 length=53',\n",
       "  'dbxrefs': [],\n",
       "  'annotations': {},\n",
       "  '_per_letter_annotations': {},\n",
       "  'features': []}]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_df.head(1).to_dict('records')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GenBank to Pandas DataFrame example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_seq</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>dbxrefs</th>\n",
       "      <th>_per_letter_annotations</th>\n",
       "      <th>molecule_type</th>\n",
       "      <th>topology</th>\n",
       "      <th>data_file_division</th>\n",
       "      <th>date</th>\n",
       "      <th>...</th>\n",
       "      <th>codon_start</th>\n",
       "      <th>transl_table</th>\n",
       "      <th>product</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>translation</th>\n",
       "      <th>EC_number</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>gene_synonym</th>\n",
       "      <th>experiment</th>\n",
       "      <th>pseudo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>NC_007795.1</td>\n",
       "      <td>NC_007795</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>{}</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>CON</td>\n",
       "      <td>03-AUG-2016</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>NC_007795.1</td>\n",
       "      <td>NC_007795</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>{}</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>CON</td>\n",
       "      <td>03-AUG-2016</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>NC_007795.1</td>\n",
       "      <td>NC_007795</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>{}</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>CON</td>\n",
       "      <td>03-AUG-2016</td>\n",
       "      <td>...</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[chromosomal replication initiation protein]</td>\n",
       "      <td>[YP_498609.1]</td>\n",
       "      <td>[MSEKEIWEKVLEIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSS...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                _seq           id       name  \\\n",
       "0  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...  NC_007795.1  NC_007795   \n",
       "1  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...  NC_007795.1  NC_007795   \n",
       "2  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...  NC_007795.1  NC_007795   \n",
       "\n",
       "                                         description  \\\n",
       "0  Staphylococcus aureus subsp. aureus NCTC 8325 ...   \n",
       "1  Staphylococcus aureus subsp. aureus NCTC 8325 ...   \n",
       "2  Staphylococcus aureus subsp. aureus NCTC 8325 ...   \n",
       "\n",
       "                                             dbxrefs _per_letter_annotations  \\\n",
       "0  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]                      {}   \n",
       "1  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]                      {}   \n",
       "2  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]                      {}   \n",
       "\n",
       "  molecule_type  topology data_file_division         date  ... codon_start  \\\n",
       "0           DNA  circular                CON  03-AUG-2016  ...         NaN   \n",
       "1           DNA  circular                CON  03-AUG-2016  ...         NaN   \n",
       "2           DNA  circular                CON  03-AUG-2016  ...         [1]   \n",
       "\n",
       "   transl_table                                       product     protein_id  \\\n",
       "0           NaN                                           NaN            NaN   \n",
       "1           NaN                                           NaN            NaN   \n",
       "2          [11]  [chromosomal replication initiation protein]  [YP_498609.1]   \n",
       "\n",
       "                                         translation EC_number pseudogene  \\\n",
       "0                                                NaN       NaN        NaN   \n",
       "1                                                NaN       NaN        NaN   \n",
       "2  [MSEKEIWEKVLEIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSS...       NaN        NaN   \n",
       "\n",
       "  gene_synonym experiment pseudo  \n",
       "0          NaN        NaN    NaN  \n",
       "1          NaN        NaN    NaN  \n",
       "2          NaN        NaN    NaN  \n",
       "\n",
       "[3 rows x 38 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gbk_df = spd.read_seq(gbk, format='genbank')\n",
    "gbk_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# See all the possible columns of meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['_seq', 'id', 'name', 'description', 'dbxrefs',\n",
       "       '_per_letter_annotations', 'molecule_type', 'topology',\n",
       "       'data_file_division', 'date', 'accessions', 'sequence_version',\n",
       "       'keywords', 'source', 'organism', 'taxonomy', 'references', 'comment',\n",
       "       'contig', 'location', 'type', 'mol_type', 'strain', 'sub_species',\n",
       "       'db_xref', 'gene', 'locus_tag', 'note', 'codon_start', 'transl_table',\n",
       "       'product', 'protein_id', 'translation', 'EC_number', 'pseudogene',\n",
       "       'gene_synonym', 'experiment', 'pseudo'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gbk_df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Seach for only rows of type CDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genome has CDS count: (2767, 38)\n",
      "Genome has GENE count: (2872, 38)\n"
     ]
    }
   ],
   "source": [
    "cds = gbk_df[gbk_df.type == 'CDS']\n",
    "gene = gbk_df[gbk_df.type == 'gene']\n",
    "print('Genome has CDS count:', cds.shape)\n",
    "print('Genome has GENE count:', gene.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_seq</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>dbxrefs</th>\n",
       "      <th>_per_letter_annotations</th>\n",
       "      <th>molecule_type</th>\n",
       "      <th>topology</th>\n",
       "      <th>data_file_division</th>\n",
       "      <th>date</th>\n",
       "      <th>...</th>\n",
       "      <th>codon_start</th>\n",
       "      <th>transl_table</th>\n",
       "      <th>product</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>translation</th>\n",
       "      <th>EC_number</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>gene_synonym</th>\n",
       "      <th>experiment</th>\n",
       "      <th>pseudo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>NC_007795.1</td>\n",
       "      <td>NC_007795</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>{}</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>CON</td>\n",
       "      <td>03-AUG-2016</td>\n",
       "      <td>...</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[chromosomal replication initiation protein]</td>\n",
       "      <td>[YP_498609.1]</td>\n",
       "      <td>[MSEKEIWEKVLEIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSS...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>NC_007795.1</td>\n",
       "      <td>NC_007795</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>{}</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>CON</td>\n",
       "      <td>03-AUG-2016</td>\n",
       "      <td>...</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[DNA polymerase III subunit beta]</td>\n",
       "      <td>[YP_498610.1]</td>\n",
       "      <td>[MMEFTIKRDYFITQLNDTLKAISPRTTLPILTGIKIDAKEHEVIL...</td>\n",
       "      <td>[2.7.7.7]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>NC_007795.1</td>\n",
       "      <td>NC_007795</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>{}</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>CON</td>\n",
       "      <td>03-AUG-2016</td>\n",
       "      <td>...</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[hypothetical protein]</td>\n",
       "      <td>[YP_498611.1]</td>\n",
       "      <td>[MIILVQEVVVEGDINLGQFLKTEGIIESGGQAKWFLQDVEVLING...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                _seq           id       name  \\\n",
       "2  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...  NC_007795.1  NC_007795   \n",
       "4  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...  NC_007795.1  NC_007795   \n",
       "6  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...  NC_007795.1  NC_007795   \n",
       "\n",
       "                                         description  \\\n",
       "2  Staphylococcus aureus subsp. aureus NCTC 8325 ...   \n",
       "4  Staphylococcus aureus subsp. aureus NCTC 8325 ...   \n",
       "6  Staphylococcus aureus subsp. aureus NCTC 8325 ...   \n",
       "\n",
       "                                             dbxrefs _per_letter_annotations  \\\n",
       "2  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]                      {}   \n",
       "4  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]                      {}   \n",
       "6  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]                      {}   \n",
       "\n",
       "  molecule_type  topology data_file_division         date  ... codon_start  \\\n",
       "2           DNA  circular                CON  03-AUG-2016  ...         [1]   \n",
       "4           DNA  circular                CON  03-AUG-2016  ...         [1]   \n",
       "6           DNA  circular                CON  03-AUG-2016  ...         [1]   \n",
       "\n",
       "   transl_table                                       product     protein_id  \\\n",
       "2          [11]  [chromosomal replication initiation protein]  [YP_498609.1]   \n",
       "4          [11]             [DNA polymerase III subunit beta]  [YP_498610.1]   \n",
       "6          [11]                        [hypothetical protein]  [YP_498611.1]   \n",
       "\n",
       "                                         translation  EC_number pseudogene  \\\n",
       "2  [MSEKEIWEKVLEIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSS...        NaN        NaN   \n",
       "4  [MMEFTIKRDYFITQLNDTLKAISPRTTLPILTGIKIDAKEHEVIL...  [2.7.7.7]        NaN   \n",
       "6  [MIILVQEVVVEGDINLGQFLKTEGIIESGGQAKWFLQDVEVLING...        NaN        NaN   \n",
       "\n",
       "  gene_synonym experiment pseudo  \n",
       "2          NaN        NaN    NaN  \n",
       "4          NaN        NaN    NaN  \n",
       "6          NaN        NaN    NaN  \n",
       "\n",
       "[3 rows x 38 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Better view of the layout for a single row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'_seq': Seq('CGATTAAAGATAGAAATACACGATGCGAGCAATCAAATTTCATAACATCACCAT...TAT'),\n",
       "  'id': 'NC_007795.1',\n",
       "  'name': 'NC_007795',\n",
       "  'description': 'Staphylococcus aureus subsp. aureus NCTC 8325 chromosome, complete genome',\n",
       "  'dbxrefs': ['BioProject:PRJNA57795', 'Assembly:GCF_000013425.1'],\n",
       "  '_per_letter_annotations': {},\n",
       "  'molecule_type': 'DNA',\n",
       "  'topology': 'circular',\n",
       "  'data_file_division': 'CON',\n",
       "  'date': '03-AUG-2016',\n",
       "  'accessions': ['NC_007795'],\n",
       "  'sequence_version': 1,\n",
       "  'keywords': ['RefSeq'],\n",
       "  'source': 'Staphylococcus aureus subsp. aureus NCTC 8325',\n",
       "  'organism': 'Staphylococcus aureus subsp. aureus NCTC 8325',\n",
       "  'taxonomy': ['Bacteria',\n",
       "   'Firmicutes',\n",
       "   'Bacilli',\n",
       "   'Bacillales',\n",
       "   'Staphylococcaceae',\n",
       "   'Staphylococcus'],\n",
       "  'references': [Reference(title='The Staphylococcus aureus NCTC8325 Genome', ...),\n",
       "   Reference(title='Direct Submission', ...),\n",
       "   Reference(title='Direct Submission', ...)],\n",
       "  'comment': 'REVIEWED REFSEQ: This record has been curated by NCBI staff. The\\nreference sequence was derived from CP000253.\\nRefSeq Category: Reference Genome\\n            UPR: UniProt Genome\\nStaphylococcus aureus subsp. aureus NCTC 8325 is available from\\nwww.narsa.net.\\nCOMPLETENESS: full length.',\n",
       "  'contig': 'join(CP000253.1:1..2821361)',\n",
       "  'location': FeatureLocation(ExactPosition(516), ExactPosition(1878), strand=1),\n",
       "  'type': 'CDS',\n",
       "  'mol_type': nan,\n",
       "  'strain': nan,\n",
       "  'sub_species': nan,\n",
       "  'db_xref': ['GeneID:3919798'],\n",
       "  'gene': ['dnaA'],\n",
       "  'locus_tag': ['SAOUHSC_00001'],\n",
       "  'note': ['binds to the dnaA-box as an ATP-bound complex at the origin of replication during the initiation of chromosomal replication; can also affect transcription of multiple genes including itself.'],\n",
       "  'codon_start': ['1'],\n",
       "  'transl_table': ['11'],\n",
       "  'product': ['chromosomal replication initiation protein'],\n",
       "  'protein_id': ['YP_498609.1'],\n",
       "  'translation': ['MSEKEIWEKVLEIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSSIPFNANWLNQQYAEIIQAILFDVVGYEVKPHFITTEELANYSNNETATPKETTKPSTETTEDNHVLGREQFNAHNTFDTFVIGPGNRFPHAASLAVAEAPAKAYNPLFIYGGVGLGKTHLMHAIGHHVLDNNPDAKVIYTSSEKFTNEFIKSIRDNEGEAFRERYRNIDVLLIDDIQFIQNKVQTQEEFFYTFNELHQNNKQIVISSDRPPKEIAQLEDRLRSRFEWGLIVDITPPDYETRMAILQKKIEEEKLDIPPEALNYIANQIQSNIRELEGALTRLLAYSQLLGKPITTELTAEALKDIIQAPKSKKITIQDIQKIVGQYYNVRIEDFSAKKRTKSIAYPRQIAMYLSRELTDFSLPKIGEEFGGRDHTTVIHAHEKISKDLKEDPIFKQEVENLEKEIRNV'],\n",
       "  'EC_number': nan,\n",
       "  'pseudogene': nan,\n",
       "  'gene_synonym': nan,\n",
       "  'experiment': nan,\n",
       "  'pseudo': nan}]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds.head(1).to_dict('records')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Machine Learning Example Application"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For simplicity of this example we will pretend we have a large sample set, but really we just using the single genbank genome. The follow will create the feature table we can use for a simple Sequential model.\n",
    "1. Sample the first 4 genes\n",
    "2. Given them random mutation value from 0 to 20\n",
    "3. Give each mutation a random \"observed in healthy patient\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['dnaA', 'recF', 'rplI', 'argJ']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sample for the first 4 genes\n",
    "genes = [gene[0] for gene in gbk_df[(gbk_df.type == 'gene') & ~gbk_df.gene.isna()]['gene'][:4].values]\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dnaA</th>\n",
       "      <th>recF</th>\n",
       "      <th>rplI</th>\n",
       "      <th>argJ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5.470563</td>\n",
       "      <td>19.242406</td>\n",
       "      <td>1.763627</td>\n",
       "      <td>17.773504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.525011</td>\n",
       "      <td>2.298095</td>\n",
       "      <td>1.344254</td>\n",
       "      <td>1.815636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19.226728</td>\n",
       "      <td>2.126331</td>\n",
       "      <td>16.821579</td>\n",
       "      <td>13.467762</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        dnaA       recF       rplI       argJ\n",
       "0   5.470563  19.242406   1.763627  17.773504\n",
       "1   0.525011   2.298095   1.344254   1.815636\n",
       "2  19.226728   2.126331  16.821579  13.467762"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create fake observed mutation rates for target genomes compared to a reference genome\n",
    "mutation_array = lambda : list(np.random.uniform(low=0, high=20, size=(1, 100))[0])\n",
    "observed_gene_mutation_per_genome = spd.BioDataFrame.from_dict({gene:mutation_array() for gene in genes})\n",
    "observed_gene_mutation_per_genome.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dnaA</th>\n",
       "      <th>recF</th>\n",
       "      <th>rplI</th>\n",
       "      <th>argJ</th>\n",
       "      <th>fake_disease</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5.470563</td>\n",
       "      <td>19.242406</td>\n",
       "      <td>1.763627</td>\n",
       "      <td>17.773504</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.525011</td>\n",
       "      <td>2.298095</td>\n",
       "      <td>1.344254</td>\n",
       "      <td>1.815636</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19.226728</td>\n",
       "      <td>2.126331</td>\n",
       "      <td>16.821579</td>\n",
       "      <td>13.467762</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19.314712</td>\n",
       "      <td>1.957057</td>\n",
       "      <td>18.395884</td>\n",
       "      <td>16.989597</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.778428</td>\n",
       "      <td>19.743984</td>\n",
       "      <td>19.671500</td>\n",
       "      <td>14.353111</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>15.141159</td>\n",
       "      <td>14.376695</td>\n",
       "      <td>4.884112</td>\n",
       "      <td>12.563653</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>19.045423</td>\n",
       "      <td>17.280185</td>\n",
       "      <td>17.711694</td>\n",
       "      <td>7.258818</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4.588786</td>\n",
       "      <td>4.565258</td>\n",
       "      <td>16.608461</td>\n",
       "      <td>0.441742</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.542041</td>\n",
       "      <td>2.096888</td>\n",
       "      <td>18.780869</td>\n",
       "      <td>14.548252</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3.525264</td>\n",
       "      <td>7.665830</td>\n",
       "      <td>1.038490</td>\n",
       "      <td>8.947828</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        dnaA       recF       rplI       argJ  fake_disease\n",
       "0   5.470563  19.242406   1.763627  17.773504           1.0\n",
       "1   0.525011   2.298095   1.344254   1.815636           0.0\n",
       "2  19.226728   2.126331  16.821579  13.467762           1.0\n",
       "3  19.314712   1.957057  18.395884  16.989597           0.0\n",
       "4   5.778428  19.743984  19.671500  14.353111           0.0\n",
       "5  15.141159  14.376695   4.884112  12.563653           0.0\n",
       "6  19.045423  17.280185  17.711694   7.258818           0.0\n",
       "7   4.588786   4.565258  16.608461   0.441742           0.0\n",
       "8   0.542041   2.096888  18.780869  14.548252           1.0\n",
       "9   3.525264   7.665830   1.038490   8.947828           0.0"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Give a 0 or 1 to show if a patient has the fake disease\n",
    "# We a biased 30 percent with disease  \n",
    "patients = np.zeros(100)\n",
    "patients[:30] = 1\n",
    "np.random.shuffle(patients)\n",
    "\n",
    "observed_gene_mutation_per_genome['fake_disease'] = list(patients)\n",
    "observed_gene_mutation_per_genome.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# optional downloads\n",
    "```bash\n",
    "pip install scikit-learn\n",
    "pip install tensorflow\n",
    "pip install matplotlib\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tensorflow.keras.models import Sequential\n",
    "from tensorflow.keras.layers import Dense \n",
    "from tensorflow.keras.callbacks import EarlyStopping\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import confusion_matrix, recall_score, precision_score\n",
    "\n",
    "import pandas as pd  # tensorflow is picky with core Pandas naming scheme"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = observed_gene_mutation_per_genome\n",
    "X_train, X_test, y_train, y_test = train_test_split(df.iloc[:,:4], df['fake_disease'],\n",
    "                                                    test_size = 0.33, random_state = 100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create general sequential model with 4 layers  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = Sequential()\n",
    "\n",
    "model.add(Dense(4, activation = 'relu')) # 4 for just random guessing\n",
    "model.add(Dense(4, activation = 'relu'))\n",
    "model.add(Dense(4, activation = 'relu'))\n",
    "model.add(Dense(1)) # just want healthy status; hence final layer is 1\n",
    "\n",
    "model.compile(optimizer='rmsprop', loss='mse') # mean squared error for sanity checking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 3.3702\n",
      "Epoch 2/100\n",
      "3/3 [==============================] - 0s 2ms/step - loss: 2.6729\n",
      "Epoch 3/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 2.1922\n",
      "Epoch 4/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 1.9241\n",
      "Epoch 5/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 1.7955\n",
      "Epoch 6/100\n",
      "3/3 [==============================] - 0s 950us/step - loss: 1.6933\n",
      "Epoch 7/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 1.5777\n",
      "Epoch 8/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 1.4495\n",
      "Epoch 9/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 1.3545\n",
      "Epoch 10/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 1.2412\n",
      "Epoch 11/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 1.1382\n",
      "Epoch 12/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 1.0580\n",
      "Epoch 13/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.9680\n",
      "Epoch 14/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.8849\n",
      "Epoch 15/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.7573\n",
      "Epoch 16/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.6909\n",
      "Epoch 17/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.6566\n",
      "Epoch 18/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.6217\n",
      "Epoch 19/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.5499\n",
      "Epoch 20/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.5210\n",
      "Epoch 21/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.4971\n",
      "Epoch 22/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.4768\n",
      "Epoch 23/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.4711\n",
      "Epoch 24/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.4571\n",
      "Epoch 25/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.4158\n",
      "Epoch 26/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.4006\n",
      "Epoch 27/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.3920\n",
      "Epoch 28/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.3889\n",
      "Epoch 29/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.3788\n",
      "Epoch 30/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.3766\n",
      "Epoch 31/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.3710\n",
      "Epoch 32/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.3624\n",
      "Epoch 33/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.3527\n",
      "Epoch 34/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.3458\n",
      "Epoch 35/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.3379\n",
      "Epoch 36/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.3298\n",
      "Epoch 37/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.3220\n",
      "Epoch 38/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.3086\n",
      "Epoch 39/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.3041\n",
      "Epoch 40/100\n",
      "3/3 [==============================] - 0s 994us/step - loss: 0.3007\n",
      "Epoch 41/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2932\n",
      "Epoch 42/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2878\n",
      "Epoch 43/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2842\n",
      "Epoch 44/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2765\n",
      "Epoch 45/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2635\n",
      "Epoch 46/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2579\n",
      "Epoch 47/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2568\n",
      "Epoch 48/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2531\n",
      "Epoch 49/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2500\n",
      "Epoch 50/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2483\n",
      "Epoch 51/100\n",
      "3/3 [==============================] - 0s 2ms/step - loss: 0.2430\n",
      "Epoch 52/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2412\n",
      "Epoch 53/100\n",
      "3/3 [==============================] - 0s 2ms/step - loss: 0.2392\n",
      "Epoch 54/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2373\n",
      "Epoch 55/100\n",
      "3/3 [==============================] - 0s 2ms/step - loss: 0.2360\n",
      "Epoch 56/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2333\n",
      "Epoch 57/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2338\n",
      "Epoch 58/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2336\n",
      "Epoch 59/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2339\n",
      "Epoch 60/100\n",
      "3/3 [==============================] - 0s 2ms/step - loss: 0.2335\n",
      "Epoch 61/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2287\n",
      "Epoch 62/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2276\n",
      "Epoch 63/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2257\n",
      "Epoch 64/100\n",
      "3/3 [==============================] - 0s 2ms/step - loss: 0.2270\n",
      "Epoch 65/100\n",
      "3/3 [==============================] - 0s 2ms/step - loss: 0.2259\n",
      "Epoch 66/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2272\n",
      "Epoch 67/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2288\n",
      "Epoch 68/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2286\n",
      "Epoch 69/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2258\n",
      "Epoch 70/100\n",
      "3/3 [==============================] - 0s 2ms/step - loss: 0.2222\n",
      "Epoch 71/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2218\n",
      "Epoch 72/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2224\n",
      "Epoch 73/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2222\n",
      "Epoch 74/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2228\n",
      "Epoch 75/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2209\n",
      "Epoch 76/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2212\n",
      "Epoch 77/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2213\n",
      "Epoch 78/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2230\n",
      "Epoch 79/100\n",
      "3/3 [==============================] - 0s 963us/step - loss: 0.2193\n",
      "Epoch 80/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2190\n",
      "Epoch 81/100\n",
      "3/3 [==============================] - 0s 2ms/step - loss: 0.2191\n",
      "Epoch 82/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2193\n",
      "Epoch 83/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2209\n",
      "Epoch 84/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2198\n",
      "Epoch 85/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2176\n",
      "Epoch 86/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2177\n",
      "Epoch 87/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2176\n",
      "Epoch 88/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2174\n",
      "Epoch 89/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2180\n",
      "Epoch 90/100\n",
      "3/3 [==============================] - 0s 2ms/step - loss: 0.2178\n",
      "Epoch 91/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2177\n",
      "Epoch 92/100\n",
      "3/3 [==============================] - 0s 2ms/step - loss: 0.2164\n",
      "Epoch 93/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2174\n",
      "Epoch 94/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2200\n",
      "Epoch 95/100\n",
      "3/3 [==============================] - 0s 2ms/step - loss: 0.2170\n",
      "Epoch 96/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2168\n",
      "Epoch 97/100\n",
      "3/3 [==============================] - 0s 2ms/step - loss: 0.2153\n",
      "Epoch 98/100\n",
      "3/3 [==============================] - 0s 2ms/step - loss: 0.2176\n",
      "Epoch 99/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2170\n",
      "Epoch 100/100\n",
      "3/3 [==============================] - 0s 1ms/step - loss: 0.2148\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<tensorflow.python.keras.callbacks.History at 0x15aa77940>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "early_stop = EarlyStopping(monitor = 'val_loss', patience = 1)\n",
    "model.fit(\n",
    "    x = X_train.to_numpy().astype('float32'), # float32 to save memory and runtime; we dont need 64 bits anyway\n",
    "    y = y_train.to_numpy().astype('float32'), \n",
    "    verbose = 1, # 0 if you dont want to see the print out\n",
    "    epochs = 100, # number of training intervals \n",
    "    # callbacks = [early_stop], # cant used validation set to have an educated stop due to data set being too small :(\n",
    "    # validation_data=(X_test.astype('float32'), y_test.astype('float32')), # data too small to use as a validator\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check how well your model was built"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAD7CAYAAAB68m/qAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAg20lEQVR4nO3deZhcdZ3v8fe3tq7eO0un00lnISSTBQIBOgyMEhkdBaIjcEEecFRwy+iMiHN9nHHwPjOj9z7jDDjidVAR0TE4I4KKTuDihmQmoGwNJJAVQhaSkKWz9L7V8r1/VHXodLrpTlLdJ1X1eT1PPamq8+s+35OTfOpXv/M755i7IyIi+S8UdAEiIpIbCnQRkQKhQBcRKRAKdBGRAqFAFxEpEAp0EZECMWKgm1nczJ4xs3VmtsHMvjhEm5vMrNnM1mYfHxubckVEZDiRUbTpBd7u7h1mFgWeMLNfuPtTg9rd7+6fyn2JIiIyGiMGumfOPOrIvoxmH6d8NtLkyZN99uzZp/prRESKynPPPXfQ3WuHWjaaHjpmFgaeA+YC33D3p4dodo2ZLQNeBv7K3Xe92e+cPXs2TU1No1m9iIhkmdnO4ZaN6qCou6fcfQnQAFxoZmcPavIQMNvdzwF+A6wcppAVZtZkZk3Nzc2jKl5EREbnhGa5uHsLsBq4fND7h9y9N/vyHuCCYX7+bndvdPfG2tohvzGIiMhJGs0sl1ozq8k+LwXeCWwe1KZ+wMv3AptyWKOIiIzCaMbQ64GV2XH0EPCAuz9sZl8Cmtx9FfBpM3svkAQOAzeNVcEiIgMlEgl2795NT09P0KXkVDwep6GhgWg0OuqfsaAun9vY2Og6KCoip2r79u1UVlYyadIkzCzocnLC3Tl06BDt7e2cccYZxywzs+fcvXGon9OZoiKS13p6egoqzAHMjEmTJp3wtw4FuojkvUIK834ns015F+hb9rXzlV9t4UhnX9CliIgAUFFREXQJQB4G+vaDndy5eiuvt3YHXYqIyGkl7wK9pixzxLe1KxFwJSIix3J3Pve5z3H22WezePFi7r//fgD27t3LsmXLWLJkCWeffTaPP/44qVSKm2666WjbO+6445TXP6pT/08n/YHe0q1AF5HTy4MPPsjatWtZt24dBw8eZOnSpSxbtowf/vCHXHbZZXzhC18glUrR1dXF2rVr2bNnD+vXrwegpaXllNeff4FeGgOgRT10ERnkiw9tYOPrbTn9nYumVfH3f3rWqNo+8cQT3HDDDYTDYerq6njb297Gs88+y9KlS/nIRz5CIpHgqquuYsmSJcyZM4dt27Zx88038+53v5t3vetdp1xr3g65tHTroKiI5Idly5axZs0apk+fzk033cS9997LhAkTWLduHZdeeil33XUXH/vYqd9GIu966PFomJJISGPoInKc0fakx8oll1zCt7/9bW688UYOHz7MmjVruP3229m5cycNDQ18/OMfp7e3l+eff57ly5cTi8W45pprmD9/Ph/4wAdOef15F+iQ6aVryEVETjdXX301Tz75JOeeey5mxm233cbUqVNZuXIlt99+O9FolIqKCu6991727NnDhz/8YdLpNABf/vKXT3n9eXnq/2V3rGH25DK+/cEhz34VkSKyadMmFi5cGHQZY2KobSu4U/+r1UMXETlOXgZ6TWmUVk1bFBE5Rn4GunroIiLHydNAj2naoogcFdSxwLF0MtuUl4FeXRqlJ5GmJ5EKuhQRCVg8HufQoUMFFer910OPx+Mn9HN5O20RoLU7QTwaDrgaEQlSQ0MDu3fvptBuPN9/x6ITkZ+BPuD0/7qqE/sEE5HCEo1Gj7urT7HKyyGXo6f/d2kcXUSkX14GenWprrgoIjJYXga6rokuInK8PA307Bi6pi6KiByVl4FeHgsTCZlOLhIRGWDEQDezuJk9Y2brzGyDmX1xiDYlZna/mW01s6fNbPaYVPvG+jJni2oMXUTkqNH00HuBt7v7ucAS4HIzu2hQm48CR9x9LnAH8M85rXII1aVRjaGLiAwwYqB7Rkf2ZTT7GHxK1pXAyuzznwDvMDPLWZVD0On/IiLHGtUYupmFzWwtcAD4jbs/PajJdGAXgLsngVZg0hC/Z4WZNZlZ06me1VVTqgt0iYgMNKpAd/eUuy8BGoALzezsk1mZu9/t7o3u3lhbW3syv+IoXRNdRORYJzTLxd1bgNXA5YMW7QFmAJhZBKgGDuWgvmHVlMZ0TXQRkQFGM8ul1sxqss9LgXcCmwc1WwXcmH1+LfCYj/Glz2rKonT0Jkmk0mO5GhGRvDGai3PVAyvNLEzmA+ABd3/YzL4ENLn7KuC7wA/MbCtwGLh+zCrOGnjFxckVJWO9OhGR096Ige7uLwLnDfH+3w143gO8L7elvbmj13PpUqCLiECenikKb5z+36qpiyIiQD4H+oAeuoiI5HOglynQRUQGyt9A779rkaYuiogAeRzolfEIZtCquxaJiAB5HOihkFFdqisuioj0y9tAB13PRURkoPwO9LKYeugiIll5HuhRjaGLiGTld6BrDF1E5Kj8DvSymMbQRUSy8jrQq0ujtPUkSKXH9MKOIiJ5Ia8DvaYsiju096iXLiKS94EOOv1fRATyPdCzp/8f0UwXEZH8DvRJFZlAP9ShQBcRyetAr6uKA7C/vSfgSkREgpfXgT6pPEbIYH+rAl1EJK8DPRIOMbmihP1tvUGXIiISuLwOdMgMu2jIRUSkIAJdPXQRESiIQI9zoE09dBGREQPdzGaY2Woz22hmG8zsliHaXGpmrWa2Nvv4u7Ep93h1VXEOdfbRl0yP1ypFRE5LkVG0SQKfdffnzawSeM7MfuPuGwe1e9zd35P7Et9cXVUJAM0dvUyvKR3v1YuInDZG7KG7+153fz77vB3YBEwf68JGa0p2Lvo+TV0UkSJ3QmPoZjYbOA94eojFF5vZOjP7hZmdlYviRqOuMhPoGkcXkWI3miEXAMysAvgp8Bl3bxu0+Hlglrt3mNly4OfAvCF+xwpgBcDMmTNPtuZj9A+57Fegi0iRG1UP3cyiZML8P9z9wcHL3b3N3Tuyzx8BomY2eYh2d7t7o7s31tbWnmLpGRPKYkTDxv52TV0UkeI2mlkuBnwX2OTuXx2mzdRsO8zswuzvPZTLQocTChlTKuPqoYtI0RvNkMtbgA8CL5nZ2ux7twIzAdz9LuBa4JNmlgS6gevdfdxuI1RXVcIBnVwkIkVuxEB39ycAG6HNncCduSrqRNVVxdl6oCOo1YuInBby/kxRyF7PRUMuIlLkCiLQp1SV0NaTpLsvFXQpIiKBKYhA75+Lrl66iBSzwgj0KgW6iEiBBHr25CLNRReRIlYYgV6t0/9FRAoi0CtLIpRGwxpyEZGiVhCBbma6c5GIFL2CCHTIXEZ3n3roIlLECibQdSs6ESl2hRPolZkhl3G8hIyIyGmlcAK9Kk53IkV7bzLoUkREAlEwgT4lOxddwy4iUqwKJtCnHr23qGa6iEhxKphAnzGxDIDtB3UZXREpTgUT6PXVcapLo2za1x50KSIigSiYQDczFkytZNPewfevFhEpDgUT6AAL66vYsq+ddFpTF0Wk+BRUoC+qr6KrL8XOw11BlyIiMu4KKtAX1lcBaNhFRIpSQQX6vLoKwiFToItIUSqoQI9Hw8yZXK5AF5GiNGKgm9kMM1ttZhvNbIOZ3TJEGzOzr5vZVjN70czOH5tyR7agvopNezV1UUSKz2h66Engs+6+CLgI+EszWzSozRXAvOxjBfCtnFZ5AhbWV7KnpZvW7kRQJYiIBGLEQHf3ve7+fPZ5O7AJmD6o2ZXAvZ7xFFBjZvU5r3YU+g+Mbtawi4gUmRMaQzez2cB5wNODFk0Hdg14vZvjQ39cLNJMFxEpUqMOdDOrAH4KfMbdTyotzWyFmTWZWVNzc/PJ/IoRTaksYWJ5TOPoIlJ0RhXoZhYlE+b/4e4PDtFkDzBjwOuG7HvHcPe73b3R3Rtra2tPpt7R1MrC+ko27VMPXUSKy2hmuRjwXWCTu391mGargA9lZ7tcBLS6+94c1nlCFkzNXAIgmUoHVYKIyLiLjKLNW4APAi+Z2drse7cCMwHc/S7gEWA5sBXoAj6c80pPwML6KnqTaXYc6mTulMogSxERGTcjBrq7PwHYCG0c+MtcFXWqFtZnQnz9njYFuogUjYI6U7Tf/LpKplSW8PCLrwddiojIuCnIQI+EQ1xzQQOrtzTrHqMiUjQKMtAB3ndBA6m08+ALx022EREpSAUb6HNqK2icNYEHmnaRGeIXESlsBRvoANc1zmBbcyfPv9YSdCkiImOuoAN9+Tn1lMXC/Lhp18iNRUTyXEEHekVJhOWL63lo3et09SWDLkdEZEwVdKBDZtilsy/Fwy8GduKqiMi4KPhAXzp7AgumVvKdNdtIp3VwVEQKV8EHupnxyUvP5JUDHfx6476gyxERGTMFH+gA7zlnGrMnlXHn6q2awigiBasoAj0cyvTS1+9p479fHpvrsIuIBK0oAh3g6vMamFYd5xurtwZdiojImCiaQI9FQqxYNodndxzh6W2Hgi5HRCTniibQAa6/cCaTymN85/HtQZciIpJzRRXo8WiYGy6cyW8372fX4a6gyxERyamiCnSAP7toJiEz/v2pnUGXIiKSU0UX6PXVpVx2Vh0/enYX3X2poMsREcmZogt0gBsvnk1rd4JV63StdBEpHEUZ6BeeMZEFUyv5/u936kQjESkYRRnoZsaNfzSbTXvbeHbHkaDLERHJiaIMdICrlkynKh7hBzo4KiIFYsRAN7PvmdkBM1s/zPJLzazVzNZmH3+X+zJzrzQW5urzpvOrDfto7UoEXY6IyCkbTQ/9+8DlI7R53N2XZB9fOvWyxsd1S2fQl0zznzo4KiIFYMRAd/c1wOFxqGXcnTWtmrOmVfGAblEnIgUgV2PoF5vZOjP7hZmdlaPfOS6ua5zB+j1tbHi9NehSREROSS4C/XlglrufC/wr8PPhGprZCjNrMrOm5ubT4zK2Vy6ZRiwc4sdNu4MuRUTklJxyoLt7m7t3ZJ8/AkTNbPIwbe9290Z3b6ytrT3VVedETVmMd51Vx8/X7qE3qTNHRSR/nXKgm9lUM7Ps8wuzvzOvrk97XeMMWroSPLrxQNCliIictMhIDczsPuBSYLKZ7Qb+HogCuPtdwLXAJ80sCXQD13uenX75lrmTmVYdZ+Xvd7B88VSyn08iInllxEB39xtGWH4ncGfOKgpAOGSsWDaHf3hoI7/beoi3zhtyxEhE5LRWtGeKDnbDH85kWnWcr/x6i67vIiJ5SYGeVRIJ8+l3zGPtrhYe26yxdBHJPwr0Aa65oIFZk8r4l1+/TDqtXrqI5BcF+gDRcIjP/Mk8Nu5t4xfr9wVdjojICVGgD/Lec6czb0oF//rYKxpLF5G8okAfJBwyPn7JHDbva+fJV/NqOr2IFDkF+hDeu2QaE8tj/NvvdwRdiojIqCnQhxCPhnn/hTN5dNN+XjvUFXQ5IiKjokAfxgcvnkXYjJVP7gi6FBGRUVGgD6OuKs7yxfU88OwuOnqTQZcjIjIiBfqb+PBbZtPem+Snz+nSuiJy+lOgv4nzZk5gyYwa/u1320npRCMROc0p0EfwibfNYcehLn6xfm/QpYiIvCkF+gjetWgqZ9aW843Vr+pEIxE5rSnQRxAKGZ+8dC6b9rbxX1tOj9vmiYgMRYE+Clcumcb0mlLuXL1VvXQROW0p0EchGg7x52+bw3M7j/DM9sNBlyMiMiQF+ihd1ziDyRUx7ly9NehSRESGpEAfpXg0zCfediaPv3KQ32zcH3Q5IiLHUaCfgBv/aDbz6yr5h1Ub6OrT2aMicnpRoJ+AaDjE/7n6bPa0dPP132roRUROLwr0E7R09kTed0ED9zy+jZf3twddjojIUQr0k/D5KxZQXhLhf/18vaYxishpY8RAN7PvmdkBM1s/zHIzs6+b2VYze9HMzs99maeXSRUlfO6y+Tyz/TC/2qADpCJyehhND/37wOVvsvwKYF72sQL41qmXdfq7fukMzqwt57ZfbiaRSgddjojIyIHu7muANzub5krgXs94Cqgxs/pcFXi6ioRD/M3lC9h2sJP7n90VdDkiIjkZQ58ODEy03dn3Ct47F9WxdPYEvvboK3TqJhgiErBxPShqZivMrMnMmpqb8/9CV2bG3y5fyMGOXu55fHvQ5YhIkctFoO8BZgx43ZB97zjufre7N7p7Y21tbQ5WHbzzZ07girOn8u01r7K/rSfockSkiOUi0FcBH8rOdrkIaHX3orobxOevWEAy5fzTLzYHXYqIFLHRTFu8D3gSmG9mu83so2b2CTP7RLbJI8A2YCvwHeAvxqza09SsSeV8fNkZ/OyFPTTt0NUYRSQYFtSJMY2Njd7U1BTIusdCV1+Sd/zLfzOhLMZDN7+VcMiCLklECpCZPefujUMt05miOVIWi3Dr8oVs3NvGfc+8FnQ5IlKEFOg59J5z6rlozkS+8usttPUkgi5HRIqMAj2HzIy/uXwBLV0JVq19PehyRKTIKNBzbMmMGhZMreTHTTp7VETGlwI9x8yM9zXOYN3uVjbvawu6HBEpIgr0MXD1edOJho0Hnt0ddCkiUkQU6GNgYnmMdy6q42cv7KYvqSsxisj4UKCPkesaZ3CkK8Gjm3S9dBEZHwr0MXLJvFqmVsV5QAdHRWScKNDHSDhkXHtBA2tebmb7wc6gyxGRIqBAH0MfuGgWFSURbvnRCxpLF5Exp0AfQ1Or49x27Tm8uLuV236pKzGKyNhSoI+xy8+u50MXz+KeJ7bz2GYdIBWRsaNAHwe3Ll/IovoqPvvAOjbt1clGIjI2FOjjIB4Nc+f7zyMSDnHlnb/j7jWvkk4Hc9liESlcCvRxMqe2gl99Zhl/vKCWf3xkM++/5yn2tHQHXZaIFBAF+jiaWB7jrg9cwO3XnsNLu1u5/GtreGidrsooIrmhQB9n/RfveuSWS5g7pYKb73uB/3n/Wjp7k0GXJiJ5ToEekFmTyvnxn1/MLe+Yx8/X7uHWn70UdEkikuciQRdQzCLhEH/1zj8gZMYdj77MZWdNZfni+qDLEpE8pR76aeAv/vhMzmmo5gs/e4kD7T1BlyMieUqBfhqIhkN89bpz6exLceuDL+GuKY0icuJGFehmdrmZbTGzrWb2+SGW32RmzWa2Nvv4WO5LLWxzp1Ty15fN59FNB/jHRzbR1aeDpCJyYkYcQzezMPAN4J3AbuBZM1vl7hsHNb3f3T81BjUWjY+85Qxe3t/Odx7fzv97cS+3vnsh715cj5kFXZqI5IHR9NAvBLa6+zZ37wN+BFw5tmUVp1DIuO3ac/nxJy6mpizGp374Ald/8/c8tnm/hmFEZESjCfTpwMC7NOzOvjfYNWb2opn9xMxm5KS6IrV09kQeuvmt/NP/WExzey8f+X4Tf3rnEzy07nVdhldEhpWraYsPAfe5e6+Z/TmwEnj74EZmtgJYATBz5swcrbowhUPG9RfO5JoLGvjZC3v45uqt3HzfC9RWlnD90hm8a9FUpk8oZUJZVEMyIgKAjfRV3swuBv7B3S/Lvv5bAHf/8jDtw8Bhd69+s9/b2NjoTU1NJ1V0MUqnnf9+pZl/f3Inj205QP9ui0dD/EFdJcvm1XLp/FqWzKghEtbkJZFCZWbPuXvjkMtGEegR4GXgHcAe4Fng/e6+YUCbenffm31+NfA37n7Rm/1eBfrJ29PSzUu7W3m9pZvXW7pZu6uF5187QtrhjMnlfOdDFzB3SmXQZYrIGHizQB9xyMXdk2b2KeBXQBj4nrtvMLMvAU3uvgr4tJm9F0gCh4Gbcla9HGd6TSnTa0qPea+1K8F/vXyA//3wRq7+5u/55p+dzyXzagOqUESCMGIPfayohz42dh/p4mMrm3jlQAefu2w+1zXOYGJ5LOiyRCRHTmnIZawo0MdOR2+SW+57gd9uPkDIoHHWRC5dUMui+ioW1lcxpbJEB1JF8tQpDblI/qkoiXDPjY2s39PGbzbu49cb93PbL7ccXV5ZEqGmPEpVPMrE8hiL6qtY3FDNWdOqmVAWpSwWIRbRgVWRfKMeepE40tnH5n3tbNnXxo5DXbR2J2jrTnCgvZct+9rpSx07vz0WDlFbWcK0mjjTako5s7aC+VMrmV9XycyJZYRC6uGLBEE9dGFCeYyLz5zExWdOOm5ZXzLNln3tbNrXRkdPkq6+JO29Sfa39vB6aw9NO47wn2vfuLNSRUmEs6ZVsXh6NfPqKpg1qZzZk8qpq9JQjkiQFOhCLBJicUM1ixuGP3WgszfJKwc62Ly3jfWvt7J+Txs/eGonvQPOXC2LhZk7pYK5tRXMratg3pRK/qCugoYJZYTVoxcZcwp0GZXykghLZtSwZEbN0feSqTR7W3vYcaiTHQc7ebW5k1ebO/j9q4d48IU9R9uVREKcMbmcM6dUcNa0Ki6ZW8tZ06o0bCOSYxpDlzHR1pNg64EOXtnfztYDHbza3MnWAx28drgLyNww+/yZNUypijO5ooTJFTFqymJUl0aZVB5j9uRyKkrU3xAZTGPoMu6q4lHOnzmB82dOOOb95vZefrf1IGtebmbD62288FoLh7v6GKpfMaWyhNmTy5leU8rU6jhTq+JMKI8xsSzGhPIosyYp9EUG0v8GGVe1lSVcdd50rjrvjQt2JlNpDnf10dadoKUrQXN7L9sPdbKtuZOdhzp5Zvth9rf1kEwfn/r11XHmTqngglkTeOvcyZw7o4aormUjRUpDLpIX0mnnUGcfLV19HOlKcKijl20HO3n1QAebszN03KE8Fmb25HLqq0uZVhOnoiRCNBwiFglRURKhpizKhLIYNWWZefjVpVHKSsLEwiHN0JG8oCEXyXuhkFFbWUJtZcmQy1u6+njy1UM8te0QOw93setwF89sP0R3IkUiNXKnxQzikTBVpRGmVsWpq4ozqaKE0miY0liI0miYeDRMaSyceS8aJh4LE4+EiUVClGQfFfEI5SURymOZ/1ppd9LZTpNhmIE7OI57Zr0hM0Jmmgkkp0yBLgWhpizGFYvruWJx/XHL0mmnL5Wmozd5tIff2pWgtTvz6E6k6Emk6E2mOdLZx762HrYf7OT5147Qk0jTnUiRGmK4J9cqSiJHP7Qqs98sotkPitLsh4lZ5ryBvmQaJ3MCWCRkhEJGXzJNb3ZZbzKzPYnsCWMGhEMhKkrCVMajVMYz31b6D0RHw3b0AyftTjoNqaMfRGBm2fedlDvR8Bs1hcxwd9LOMXfWSqadnkSKnkQaM45+O6ooiRAKGSHLfMg5mZ/JfNC9sc96k2l6Ein6UmkiISMWCR1db3kskl33Gz/Tr7/ezJ+Qdkil3/hgPaZttoZwaOgP1EQqTWdvko7eJH3JNBUlEapKo5RERv5Gl0575veP4zc/BboUvFDIiIcyPezJFUP38EfSl0zTk0zR3Zeiqy+VDaoU3YnU0YDtSWb/8/ck6exLYmRCq396pvsbvXKzAb11d1JpaOnu40B7L81tvexv7yGRzHwQ9SUzHyrdfSlS7pRkgz5kkEg5iVSatDuxcIhYJHz020JJNEw03L/uTMB29iZp70nQ3pMc8phEMQuHjFg4RDhkJFLp7N/r8G3D2X0YMiMatszfd8iO/jvoP0fDDCIhIx4JU1aS+TB6/x/O5GOXzMn5NijQRUYhFsmMw1fFo0GXkhPunv3GkvmWks72sNPumeGfbFi90T4TTP092UQq03vu6kuRdo72tjM93oxwyIhnh6rS7rR0JTjS1UdHT/LokFO6/wMu+zP9PWYsc/5CPJo5vpFKZz64+lJpuvtSdPal6O5LHp0d1V9rfy9/4LBWZjgr88E6+BtBv2TK6UtlPpxTaYhGMuEeC4coL4lkjsVEjI7eFO09CTp7k0f/vtJpJ5FyepMp+pJOaSxEeSxCPBoGMm36v6109aboSqROumMxEgW6SBEys+zQSxTdALhwaH6XiEiBUKCLiBQIBbqISIFQoIuIFAgFuohIgVCgi4gUCAW6iEiBUKCLiBSIwK62aGbNwM6T/PHJwMEclpMvinG7i3GboTi3uxi3GU58u2e5e+1QCwIL9FNhZk3DXT6ykBXjdhfjNkNxbncxbjPkdrs15CIiUiAU6CIiBSJfA/3uoAsISDFudzFuMxTndhfjNkMOtzsvx9BFROR4+dpDFxGRQfIu0M3scjPbYmZbzezzQdczFsxshpmtNrONZrbBzG7Jvj/RzH5jZq9k/5wQdK1jwczCZvaCmT2cfX2GmT2d3ef3m1ks6BpzycxqzOwnZrbZzDaZ2cXFsK/N7K+y/77Xm9l9ZhYvxH1tZt8zswNmtn7Ae0PuX8v4enb7XzSz809kXXkV6GYWBr4BXAEsAm4ws0XBVjUmksBn3X0RcBHwl9nt/DzwW3efB/w2+7oQ3QJsGvD6n4E73H0ucAT4aCBVjZ3/C/zS3RcA55LZ9oLe12Y2Hfg00OjuZwNh4HoKc19/H7h80HvD7d8rgHnZxwrgWyeyorwKdOBCYKu7b3P3PuBHwJUB15Rz7r7X3Z/PPm8n8x98OpltXZltthK4KpACx5CZNQDvBu7Jvjbg7cBPsk0KarvNrBpYBnwXwN373L2FItjXZO6YVmpmEaAM2EsB7mt3XwMcHvT2cPv3SuBez3gKqDGz4+98Pox8C/TpwK4Br3dn3ytYZjYbOA94Gqhz973ZRfuAuqDqGkNfA/4aSGdfTwJa3D2ZfV1o+/wMoBn4t+ww0z1mVk6B72t33wN8BXiNTJC3As9R2Pt6oOH27yllXL4FelExswrgp8Bn3L1t4DLPTE8qqClKZvYe4IC7Pxd0LeMoApwPfMvdzwM6GTS8UqD7egKZ3ugZwDSgnOOHJYpCLvdvvgX6HjjmnrYN2fcKjplFyYT5f7j7g9m39/d//cr+eSCo+sbIW4D3mtkOMsNpbyczvlyT/VoOhbfPdwO73f3p7OufkAn4Qt/XfwJsd/dmd08AD5LZ/4W8rwcabv+eUsblW6A/C8zLHgmPkTmIsirgmnIuO278XWCTu391wKJVwI3Z5zcC/znetY0ld/9bd29w99lk9u1j7v5nwGrg2myzgtpud98H7DKz+dm33gFspMD3NZmhlovMrCz7771/uwt2Xw8y3P5dBXwoO9vlIqB1wNDMyNw9rx7AcuBl4FXgC0HXM0bb+FYyX8FeBNZmH8vJjCf/FngFeBSYGHStY/h3cCnwcPb5HOAZYCvwY6Ak6PpyvK1LgKbs/v45MKEY9jXwRWAzsB74AVBSiPsauI/McYIEmW9kHx1u/wJGZibfq8BLZGYBjXpdOlNURKRA5NuQi4iIDEOBLiJSIBToIiIFQoEuIlIgFOgiIgVCgS4iUiAU6CIiBUKBLiJSIP4/t336tU74hI0AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "loss = pd.DataFrame(model.history.history)\n",
    "loss.plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<tensorflow.python.keras.engine.sequential.Sequential at 0x15a8fc340>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model # Thats exciting! This is also a word of caution that you can make a model from anything. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--------"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
